packages/server/src/adapters/next-app-dir/rethrowNextErrors.ts

```diff
@@ -1,24 +1,60 @@
-import { isNotFoundError } from 'next/dist/client/components/not-found';
-import { isRedirectError } from 'next/dist/client/components/redirect';
-import {
-  notFound as __notFound,
-  redirect as __redirect,
-} from 'next/navigation';
+import * as nextNavigation from 'next/navigation.js';
 import type { TRPCError } from '../../@trpc/server';
 import { TRPCRedirectError } from './redirect';
 
 /**
- * Rethrow errors that should be handled by Next.js
+ * @remarks The helpers from `next/dist/client/components/*` has been removed in Next.js 15.
+ * Inlining them here instead...
  */
+const REDIRECT_ERROR_CODE = 'NEXT_REDIRECT';
+function isRedirectError(error: unknown) {
+  if (
+    typeof error !== 'object' ||
+    error === null ||
+    !('digest' in error) ||
+    typeof error.digest !== 'string'
+  ) {
+    return false;
+  }
+  const [errorCode, type, destination, status] = error.digest.split(';', 4);
+  const statusCode = Number(status);
+  return (
+    errorCode === REDIRECT_ERROR_CODE &&
+    (type === 'replace' || type === 'push') &&
+    typeof destination === 'string' &&
+    !isNaN(statusCode)
+  );
+}
 
+const NOT_FOUND_ERROR_CODE = 'NEXT_NOT_FOUND';
+function isNotFoundError(error: unknown) {
+  if (typeof error !== 'object' || error === null || !('digest' in error)) {
+    return false;
+  }
+  return error.digest === NOT_FOUND_ERROR_CODE;
+}
```

Comment:
Maybe we should link to where they are defined in Next.js so it'll be easier to keep them up-to-date

Reply:
Added

Reply:
although we shouldn't have to update them as newer versions of next will just use `unstable_rethrow`. they're there just as a fallback for Next 14 users